File: NextcloudTalk/NCChatFileController.m
/**
 * SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
 * SPDX-License-Identifier: GPL-3.0-or-later
 */

#import "NCChatFileController.h"

@import NextcloudKit;

#import "NCAPIController.h"
#import "NCDatabaseManager.h"

#import "NextcloudTalk-Swift.h"

NSString * const NCChatFileControllerDidChangeIsDownloadingNotification     = @"NCChatFileControllerDidChangeIsDownloadingNotification";
NSString * const NCChatFileControllerDidChangeDownloadProgressNotification  = @"NCChatFileControllerDidChangeDownloadProgressNotification";

int const kNCChatFileControllerDeleteFilesOlderThanDays = 7;

@interface NCChatFileController ()

@property (nonatomic, strong) NCChatFileStatus *fileStatus;

@end


@implementation NCChatFileController

- (void)initDownloadDirectoryForAccount:(TalkAccount *)account
{
    NSString *encodedAccountId = [account.accountId stringByAddingPercentEncodingWithAllowedCharacters:NSCharacterSet.URLHostAllowedCharacterSet];
    NSFileManager *fileManager = [NSFileManager defaultManager];
    _tempDirectoryPath = [NSTemporaryDirectory() stringByAppendingPathComponent:@"/download/"];
    _tempDirectoryPath = [_tempDirectoryPath stringByAppendingPathComponent:encodedAccountId];
    
    NSLog(@"Directory for downloads: %@", _tempDirectoryPath);
    
    if (![fileManager fileExistsAtPath:_tempDirectoryPath]) {
        // Make sure our download directory exists
        [fileManager createDirectoryAtPath:_tempDirectoryPath withIntermediateDirectories:YES attributes:nil error:nil];
    }
    
    [self removeOldFilesFromCache:kNCChatFileControllerDeleteFilesOlderThanDays];
}

- (void)removeOldFilesFromCache:(int)thresholdDays
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSDirectoryEnumerator *enumerator = [fileManager enumeratorAtPath:_tempDirectoryPath];
    
    NSDateComponents *dayComponent = [[NSDateComponents alloc] init];
    dayComponent.day = -thresholdDays;

    NSDate *thresholdDate = [[NSCalendar currentCalendar] dateByAddingComponents:dayComponent toDate:[NSDate date] options:0];
    NSString *file;
    
    while (file = [enumerator nextObject])
    {
        NSString *filePath = [_tempDirectoryPath stringByAppendingPathComponent:file];
        NSDate *creationDate = [[fileManager attributesOfItemAtPath:filePath error:nil] fileCreationDate];
        
        if ([creationDate compare:thresholdDate] == NSOrderedAscending) {
            NSLog(@"Deleting file from cache: %@", filePath);
        
            [fileManager removeItemAtPath:filePath error:nil];
        }
    }
}

- (void)deleteDownloadDirectoryForAccount:(TalkAccount *)account
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    
    [self initDownloadDirectoryForAccount:account];
    [fileManager removeItemAtPath:_tempDirectoryPath error:nil];
    
    NSLog(@"Deleted download directory: %@", _tempDirectoryPath);
}

- (void)clearDownloadDirectoryForAccount:(TalkAccount *)account
{
    [self deleteDownloadDirectoryForAccount:account];
    [self initDownloadDirectoryForAccount:account];
}

- (NSInteger)getDiskUsageForAccount:(TalkAccount *)account
{
    [self initDownloadDirectoryForAccount:account];

    NSFileManager *fileManager = [NSFileManager defaultManager];
    NSDirectoryEnumerator *enumerator = [fileManager enumeratorAtPath:_tempDirectoryPath];

    NSString *file;
    NSInteger folderSize = 0;

    while (file = [enumerator nextObject])
    {
        NSDictionary *fileAttributes = [[NSFileManager defaultManager] attributesOfItemAtPath:[_tempDirectoryPath stringByAppendingPathComponent:file] error:nil];
        folderSize += [[fileAttributes objectForKey:NSFileSize] intValue];
    }

    return folderSize;
}

- (BOOL)isFileInCache:(NSString *)filePath withModificationDate:(NSDate *)date withSize:(double)size
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    
    if (![fileManager fileExistsAtPath:filePath]) {
        return NO;
    }
    
    NSError *error = nil;
    NSDictionary<NSFileAttributeKey, id> *fileAttributes = [fileManager attributesOfItemAtPath:filePath error:&error];
    
    NSDate *modificationDate = [fileAttributes fileModificationDate];
    long long fileSize = [fileAttributes fileSize];
    
    if ([date compare:modificationDate] == NSOrderedSame && fileSize == (long long)size) {
        return YES;
    }
    
    // At this point there's a file in our cache but there's a different one on the server
    NSLog(@"Deleting file from cache: %@", filePath);
    [fileManager removeItemAtPath:filePath error:nil];
    
    return NO;
}

- (void)setCreationDateOnFile:(NSString *)filePath withCreationDate:(NSDate *)date
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    
    NSDictionary *creationDateAttr = [NSDictionary dictionaryWithObjectsAndKeys:date, NSFileCreationDate, nil];
    [fileManager setAttributes:creationDateAttr ofItemAtPath:filePath error:nil];
}

- (void)setModificationDateOnFile:(NSString *)filePath withModificationDate:(NSDate *)date
{
    NSFileManager *fileManager = [NSFileManager defaultManager];
    
    NSDictionary *modificationDateAttr = [NSDictionary dictionaryWithObjectsAndKeys:date, NSFileModificationDate, nil];
    [fileManager setAttributes:modificationDateAttr ofItemAtPath:filePath error:nil];
}

- (void)downloadFileFromMessage:(NCMessageFileParameter *)fileParameter
{
    _fileStatus = [[NCChatFileStatus alloc] initWithFileId:fileParameter.parameterId fileName:fileParameter.name filePath:fileParameter.path fileLocalPath: nil];
    fileParameter.fileStatus = _fileStatus;
    
    [self startDownload];
}

- (void)downloadFileWithFileId:(NSString *)fileId
{
    TalkAccount *activeAccount = [[NCDatabaseManager sharedInstance] activeAccount];

    [[NCAPIController sharedInstance] getFileByIdForAccount:activeAccount withFileId:fileId completionBlock:^(NKFile * _Nullable file, NKError * _Nullable error) {
        if (file) {
            NSString *remoteDavPrefix = [NSString stringWithFormat:@"/remote.php/dav/files/%@/", activeAccount.userId];
            NSString *directoryPath = [file.path componentsSeparatedByString:remoteDavPrefix].lastObject;

            NSString *filePath = [NSString stringWithFormat:@"%@%@", directoryPath, file.fileName];

            self->_fileStatus = [[NCChatFileStatus alloc] initWithFileId:file.fileId fileName:file.fileName filePath:filePath fileLocalPath:nil];
            [self startDownload];
        } else {
            NSLog(@"An error occurred while getting file with fileId %@: %@", fileId, error.errorDescription);
            [self.delegate fileControllerDidFailLoadingFile:self withFileId:fileId withErrorDescription:error.errorDescription];
        }
    }];
}

- (bool)moveFileToTemporaryDirectoryFromSourcePath:(NSString *)sourcePath destinationPath:(NSString *)destinationPath {
    TalkAccount *activeAccount = [[NCDatabaseManager sharedInstance] activeAccount];
    [self initDownloadDirectoryForAccount:activeAccount];

    if (!sourcePath || !destinationPath) {
        NSLog(@"Missing file information. File could not be moved.");
        return false;
    }

    if ([[NSFileManager defaultManager] fileExistsAtPath:destinationPath]) {
        NSLog(@"File is already in temporary directory: %@", destinationPath);
        return false;
    }

    NSError *error = nil;
    if ([[NSFileManager defaultManager] moveItemAtPath:sourcePath toPath:destinationPath error:&error]) {
        NSLog(@"File successfully moved to: %@", destinationPath);
        return true;
    } else {
        NSLog(@"Error while moving file to temporary directory: %@", error.localizedDescription);
    }

    return false;
}

- (void)startDownload
{
    TalkAccount *activeAccount = [[NCDatabaseManager sharedInstance] activeAccount];
    
    [[NCAPIController sharedInstance] setupNCCommunicationForAccount:activeAccount];
    [self initDownloadDirectoryForAccount:activeAccount];
    
    NSString *serverUrlFileName = [NSString stringWithFormat:@"%@%@/%@", activeAccount.server, [[NCAPIController sharedInstance] filesPathForAccount:activeAccount], _fileStatus.filePath];
    _fileStatus.fileLocalPath = [_tempDirectoryPath stringByAppendingPathComponent:_fileStatus.fileName];
    
    // Setting just isDownloading without a concrete progress will show an indeterminate activity indicator
    [self didChangeIsDownloadingNotification:YES];
    
    // First read metadata from the file and check if we already downloaded it
    NKRequestOptions *options = [[NKRequestOptions alloc] initWithEndpoint:nil customHeader:nil customUserAgent:nil contentType:nil e2eToken:nil timeout:60 queue:dispatch_get_main_queue()];
    [[NextcloudKit shared] readFileOrFolderWithServerUrlFileName:serverUrlFileName depth:@"0" showHiddenFiles:NO includeHiddenFiles:@[] requestBody:nil options:options completion:^(NSString *account, NSArray<NKFile *> *files, NSData *responseDates, NKError *error) {
        if (error.errorCode == 0 && files.count == 1) {
            // File exists on server -> check our cache
            NKFile *file = files.firstObject;
        
            if ([self isFileInCache:self->_fileStatus.fileLocalPath withModificationDate:file.date withSize:file.size]) {
                NSLog(@"Found file in cache: %@", self->_fileStatus.fileLocalPath);
                
                [self.delegate fileControllerDidLoadFile:self withFileStatus:self->_fileStatus];
                [self didChangeIsDownloadingNotification:NO];
                
                return;
            }
            [[NextcloudKit shared] downloadWithServerUrlFileName:serverUrlFileName fileNameLocalPath:self->_fileStatus.fileLocalPath customUserAgent:nil addCustomHeaders:nil queue:dispatch_get_main_queue() taskHandler:^(NSURLSessionTask *task) {
                NSLog(@"Download task");
            } progressHandler:^(NSProgress *progress) {
                [self didChangeDownloadProgressNotification:progress];
            } completionHandler:^(NSString *account, NSString *etag, NSDate *date, int64_t length, NSDictionary *allHeaderFields, NKError *error) {
                if (error.errorCode == 0) {
                    // Set modification date to invalidate our cache
                    [self setModificationDateOnFile:self->_fileStatus.fileLocalPath withModificationDate:file.date];

                    // Set creation date to delete older files from cache
                    [self setCreationDateOnFile:self->_fileStatus.fileLocalPath withCreationDate:[NSDate date]];

                    [self.delegate fileControllerDidLoadFile:self withFileStatus:self->_fileStatus];
                } else {
                    NSLog(@"Error downloading file: %ld - %@", error.errorCode, error.errorDescription);
                    [self.delegate fileControllerDidFailLoadingFile:self withFileId:self->_fileStatus.fileId withErrorDescription:error.errorDescription];
                }

                [self didChangeIsDownloadingNotification:NO];
            }];
        } else {
            [self didChangeIsDownloadingNotification:NO];
            
            NSLog(@"Error downloading file: %ld - %@", error.errorCode, error.errorDescription);
            [self.delegate fileControllerDidFailLoadingFile:self withFileId:self->_fileStatus.fileId withErrorDescription:error.errorDescription];
        }
    }];
}

- (void)didChangeIsDownloadingNotification:(BOOL)isDownloading
{
    _fileStatus.isDownloading = isDownloading;
    
    NSMutableDictionary *userInfo = [NSMutableDictionary new];
    [userInfo setObject:_fileStatus forKey:@"fileStatus"];
    [[NSNotificationCenter defaultCenter] postNotificationName:NCChatFileControllerDidChangeIsDownloadingNotification
                                                        object:self
                                                      userInfo:userInfo];
}

- (void)didChangeDownloadProgressNotification:(NSProgress *)progress
{
    _fileStatus.downloadProgress = progress.fractionCompleted;
    _fileStatus.canReportProgress = (progress.totalUnitCount != -1);

    NSMutableDictionary *userInfo = [NSMutableDictionary new];
    [userInfo setObject:_fileStatus forKey:@"fileStatus"];
    [[NSNotificationCenter defaultCenter] postNotificationName:NCChatFileControllerDidChangeDownloadProgressNotification
                                                        object:self
                                                      userInfo:userInfo];
}

@end

File: NextcloudTalk/Chat/NCChatFileController.swift
//
// SPDX-FileCopyrightText: 2020 Nextcloud GmbH and Nextcloud contributors
// SPDX-License-Identifier: GPL-3.0-or-later
//

import Foundation
import NextcloudKit

public extension NSNotification.Name {
    static let NCChatFileControllerDidChangeIsDownloading = NSNotification.Name("NCChatFileControllerDidChangeIsDownloadingNotification")
    static let NCChatFileControllerDidChangeDownloadProgress = NSNotification.Name("NCChatFileControllerDidChangeDownloadProgressNotification")
}

public protocol NCChatFileControllerDelegate: AnyObject {
    func fileControllerDidLoadFile(_ fileController: NCChatFileController, with fileStatus: NCChatFileStatus)
    func fileControllerDidFailLoadingFile(_ fileController: NCChatFileController, withFileId fileId: String, withErrorDescription errorDescription: String)
}

public class NCChatFileController: NSObject {

    public weak var delegate: NCChatFileControllerDelegate?

    public var messageType: String?
    public var actionType: String?
    public private(set) var tempDirectoryPath = ""

    private let account: TalkAccount
    private let deleteFilesOlderThanDays = 7
    private var fileStatus: NCChatFileStatus?
    private var cancelDownloadHandler: (() -> Void)?
    private var isCancelled = false

    init(account: TalkAccount) {
        self.account = account

        super.init()

        self.initDownloadDirectory()
        AllocationTracker.shared.addAllocation()
    }

    deinit {
        AllocationTracker.shared.removeAllocation()
    }

    private func initDownloadDirectory() {
        let encodedAccountId = self.account.accountId.addingPercentEncoding(withAllowedCharacters: .urlHostAllowed) ?? ""
        let fileManager = FileManager.default

        tempDirectoryPath = (NSTemporaryDirectory() as NSString).appendingPathComponent("download")
        tempDirectoryPath = (tempDirectoryPath as NSString).appendingPathComponent(encodedAccountId)

        print("Directory for downloads: \(tempDirectoryPath)")

        if !fileManager.fileExists(atPath: tempDirectoryPath) {
            // Make sure our download directory exists
            try? fileManager.createDirectory(atPath: tempDirectoryPath, withIntermediateDirectories: true)
        }
    }

    public func removeOldFilesFromCache() {
        let fileManager = FileManager.default

        guard let enumerator = fileManager.enumerator(atPath: tempDirectoryPath),
              let thresholdDate = Calendar.current.date(byAdding: .day, value: -deleteFilesOlderThanDays, to: Date())
        else { return }

        for case let file as String in enumerator {
            let filePath = (tempDirectoryPath as NSString).appendingPathComponent(file)
            let creationDate = (try? fileManager.attributesOfItem(atPath: filePath))?[.creationDate] as? Date

            if let creationDate, creationDate.compare(thresholdDate) == .orderedAscending {
                print("Deleting file from cache: \(filePath)")

                try? fileManager.removeItem(atPath: filePath)
            }
        }
    }

    public func deleteDownloadDirectory() {
        try? FileManager.default.removeItem(atPath: tempDirectoryPath)

        print("Deleted download directory: \(tempDirectoryPath)")
    }

    public func clearDownloadDirectory() {
        deleteDownloadDirectory()
        initDownloadDirectory()
    }

    public func getDiskUsage() -> Int {
        let fileManager = FileManager.default

        guard let enumerator = fileManager.enumerator(atPath: tempDirectoryPath) else { return 0 }

        var folderSize = 0

        for case let file as String in enumerator {
            let filePath = (tempDirectoryPath as NSString).appendingPathComponent(file)
            let fileAttributes = try? fileManager.attributesOfItem(atPath: filePath)
            folderSize += (fileAttributes?[.size] as? Int) ?? 0
        }

        return folderSize
    }

    private func isFileInCache(_ filePath: String, withModificationDate date: Date, withSize size: Int64) -> Bool {
        let fileManager = FileManager.default

        guard fileManager.fileExists(atPath: filePath) else { return false }

        let fileAttributes = try? fileManager.attributesOfItem(atPath: filePath)
        let modificationDate = fileAttributes?[.modificationDate] as? Date
        let fileSize = (fileAttributes?[.size] as? Int64) ?? 0

        if let modificationDate, date.compare(modificationDate) == .orderedSame, fileSize == size {
            return true
        }

        // At this point there's a file in our cache but there's a different one on the server
        print("Deleting file from cache: \(filePath)")
        try? fileManager.removeItem(atPath: filePath)

        return false
    }

    private func setDate(onFile filePath: String, withCreationDate creationDate: Date?, withModificationDate modificationDate: Date?) {
        // Setting both, modification- and creationDate in one go does not work, we will end up with the modification date in both fields
        // Also the creationDate needs to be set after modificationDate (most likely because creation cannot be later than modification in theory)
        if let modificationDate {
            try? FileManager.default.setAttributes([.modificationDate: modificationDate], ofItemAtPath: filePath)
        }

        if let creationDate {
            try? FileManager.default.setAttributes([.creationDate: creationDate], ofItemAtPath: filePath)
        }
    }

    public func moveFileToTemporaryDirectory(fromSourcePath sourcePath: String, destinationPath: String) -> Bool {
        let fileManager = FileManager.default

        if fileManager.fileExists(atPath: destinationPath) {
            print("File is already in temporary directory: \(destinationPath)")
            return false
        }

        do {
            try fileManager.moveItem(atPath: sourcePath, toPath: destinationPath)
            print("File successfully moved to: \(destinationPath)")
            return true
        } catch {
            print("Error while moving file to temporary directory: \(error.localizedDescription)")
            return false
        }
    }

    ///
    /// Locally cached file, without asking the server whether our copy is still current.
    ///
    /// Only matches when the size is the one announced in the chat message, so a file that was
    /// replaced with a differently sized one on the server is not returned. A replacement with the
    /// exact same size is only caught once `downloadFile(withFileId:)` has validated the file.
    ///
    public func cachedFileURL(forFileNamed fileName: String, expectedSize: Int) -> URL? {
        guard expectedSize > 0 else { return nil }

        let filePath = (self.tempDirectoryPath as NSString).appendingPathComponent(fileName)

        guard FileManager.default.fileExists(atPath: filePath),
              let attributes = try? FileManager.default.attributesOfItem(atPath: filePath),
              let size = attributes[.size] as? Int, size == expectedSize
        else { return nil }

        return URL(fileURLWithPath: filePath)
    }

    // Stops an ongoing download. No delegate method is called afterwards.
    public func cancelDownload() {
        self.isCancelled = true
        self.cancelDownloadHandler?()
        self.cancelDownloadHandler = nil

        if self.fileStatus?.isDownloading == true {
            self.didChangeIsDownloadingNotification(isDownloading: false)
        }
    }

    public func downloadFile(withFileId fileId: String) {
        self.isCancelled = false

        // getFileById already sets up NextcloudKit
        NCAPIController.sharedInstance().getFileById(forAccount: self.account, withFileId: fileId) { file, error in
            guard !self.isCancelled else { return }

            guard let file else {
                print("An error occurred while getting file with fileId \(fileId): \(error?.errorDescription ?? "")")
                self.delegate?.fileControllerDidFailLoadingFile(self, withFileId: fileId, withErrorDescription: error?.errorDescription ?? "")
                return
            }

            let remoteDavPrefix = "/remote.php/dav/files/\(self.account.userId)/"
            let directoryPath = file.path.components(separatedBy: remoteDavPrefix).last ?? ""

            let filePath = "\(directoryPath)\(file.fileName)"

            let fileStatus = NCChatFileStatus(fileId: file.fileId, fileName: file.fileName, filePath: filePath)
            self.fileStatus = fileStatus

            let serverUrlFileName = "\(self.account.server)\(NCAPIController.sharedInstance().filesPath(forAccount: self.account))/\(fileStatus.filePath)"
            let fileLocalPath = (self.tempDirectoryPath as NSString).appendingPathComponent(fileStatus.fileName)
            fileStatus.fileLocalPath = fileLocalPath

            // Setting just isDownloading without a concrete progress will show an indeterminate activity indicator
            self.didChangeIsDownloadingNotification(isDownloading: true)

            // File exists on server -> check our cache
            if self.isFileInCache(fileLocalPath, withModificationDate: file.date as Date, withSize: file.size) {
                print("Found file in cache: \(fileLocalPath)")

                self.delegate?.fileControllerDidLoadFile(self, with: fileStatus)
                self.didChangeIsDownloadingNotification(isDownloading: false)

                return
            }

            NextcloudKit.shared.download(serverUrlFileName: serverUrlFileName, fileNameLocalPath: fileLocalPath, queue: .main) { request in
                self.cancelDownloadHandler = { _ = request.cancel() }
            } progressHandler: { progress in
                self.didChangeDownloadProgressNotification(progress: progress)
            } completionHandler: { _, _, _, _, _, error in
                self.cancelDownloadHandler = nil

                guard !self.isCancelled else { return }

                if error.errorCode == 0 {
                    // Set modification date to invalidate our cache
                    // Set creation date to delete older files from cache
                    self.setDate(onFile: fileLocalPath, withCreationDate: Date(), withModificationDate: file.date as Date)

                    self.delegate?.fileControllerDidLoadFile(self, with: fileStatus)
                } else {
                    print("Error downloading file: \(error.errorCode) - \(error.errorDescription)")
                    self.delegate?.fileControllerDidFailLoadingFile(self, withFileId: fileStatus.fileId, withErrorDescription: error.errorDescription)
                }

                self.didChangeIsDownloadingNotification(isDownloading: false)
            }
        }
    }

    private func didChangeIsDownloadingNotification(isDownloading: Bool) {
        guard let fileStatus else { return }

        fileStatus.isDownloading = isDownloading

        NotificationCenter.default.post(name: .NCChatFileControllerDidChangeIsDownloading, object: self, userInfo: ["fileStatus": fileStatus])
    }

    private func didChangeDownloadProgressNotification(progress: Progress) {
        guard let fileStatus else { return }

        fileStatus.downloadProgress = Float(progress.fractionCompleted)
        fileStatus.canReportProgress = progress.totalUnitCount != -1

        NotificationCenter.default.post(name: .NCChatFileControllerDidChangeDownloadProgress, object: self, userInfo: ["fileStatus": fileStatus])
    }
}
